### PR TITLE
Pending Example Impovement

### PR DESCRIPTION
**Pull request by takecian**
This PR is related to #490 .
- print detailed description for pending example
- execute before/after hooks for pending example
- rename Filter.pending to Filter.excluded

please review, thanks.
Takeshi.

### CODE REVIEW

**Comment by takecian** [comment]
@modocache, if you have spare time, it would be helpful that you review this! :+1:

File: Sources/QuickTests/FunctionalTests/PendingTests.swift [hunk 1]
```diff
@@ -45,13 +45,13 @@ class PendingTests: XCTestCase, XCTestCaseProvider {
...
-        XCTAssertEqual(oneExampleBeforeEachExecutedCount, 1)
+        XCTAssertEqual(oneExampleBeforeEachExecutedCount, 2)
     }
 
     func testBeforeEachDoesNotRunForContextsWithOnlyPendingExamples() {
         onlyPendingExamplesBeforeEachExecutedCount = 0
 
         qck_runSpec(FunctionalTests_PendingSpec.self)
-        XCTAssertEqual(onlyPendingExamplesBeforeEachExecutedCount, 0)
+        XCTAssertEqual(onlyPendingExamplesBeforeEachExecutedCount, 1)
```
Comment (modocache) [inline Sources/QuickTests/FunctionalTests/PendingTests.swift]: This assertion no longer matches the name of the test. The test is named "test that `beforeEach` does not run for `context`s with only `pending` examples", whereas this assertion verifies that `beforeEach` **does** run. The same goes for `testBeforeEachOnlyRunForEnabledExamples` above.

So, why do these tests exist? `pending` examples were first implemented in #38. At that time, Quick had an ad hoc test suite: a single test file that used Quick and was expected to pass. I added in-depth tests in #154, which included tests for `pending` in 6b7681b6d78e8c4944fe91ad81b49267aaf59821. Unfortunately, I neglected to realize the importance of documenting in the commit message whether `beforeEach` not running was intended, or merely the behavior at the time that I wrote these tests.

I think these tests merely documented the behavior of `pending` at the time. It seems like the purpose of this PR is to ensure that `beforeEach`, `afterEach`, `beforeSuite`, and `afterSuite` always run, even for `pending` examples. If that's the case, please:

1. Delete the `testBeforeEachOnlyRunForEnabledExamples` and `testBeforeEachDoesNotRunForContextsWithOnlyPendingExamples` tests.
2. Add tests for `beforeEach`, `afterEach`, `beforeSuite`, and `afterSuite`.

**Comment by modocache** [comment]
@takecian This is awesome!! Thanks for working on this, and sorry for the wait in reviewing.

File: Sources/Quick/DSL/World+DSL.swift [hunk 2]
```diff
@@ -155,10 +155,12 @@ extension World {
...
 
-    internal func pending(description: String, closure: () -> ()) {
-        print("Pending: \(description)")
+    internal func pending(description: String, flags: FilterFlags, file: String, line: UInt, closure: () -> ()) {
+        var pendingFlags = flags
+        pendingFlags[Filter.pending] = true
+        self.it(description, flags: pendingFlags, file: file, line: line, closure: closure)
     }
-
+
```
Comment (modocache) [inline Sources/Quick/DSL/World+DSL.swift]: nit-pick: Remove this extra whitespace.

File: Sources/Quick/DSL/World+DSL.h [hunk 3]
```diff
@@ -16,5 +16,5 @@
 - (void)fitWithDescription:(NSString * __nonnull)description flags:(NSDictionary * __nonnull)flags file:(NSString * __nonnull)file line:(NSUInteger)line closure:(void (^ __nonnull)(void))closure;
 - (void)xitWithDescription:(NSString * __nonnull)description flags:(NSDictionary * __nonnull)flags file:(NSString * __nonnull)file line:(NSUInteger)line closure:(void (^ __nonnull)(void))closure;
 - (void)itBehavesLikeSharedExampleNamed:(NSString * __nonnull)name sharedExampleContext:(NSDictionary * __nonnull (^ __nonnull)(void))sharedExampleContext flags:(NSDictionary * __nonnull)flags file:(NSString * __nonnull)file line:(NSUInteger)line;
-- (void)pending:(NSString * __nonnull)description closure:(void (^ __nonnull)(void))closure;
+- (void)pending:(NSString * __nonnull)description flags:(NSDictionary * __nonnull)flags file:(NSString * __nonnull)file line:(NSUInteger)line closure:(void (^ __nonnull)(void))closure;
```
Comment (modocache) [inline Sources/Quick/DSL/World+DSL.h]: I thought it was cool how #38, which initially implemented `pending`, marked the `closure` parameter as `__unused`. Should we do that here as well?

File: Sources/Quick/Filter.swift [hunk 4]
```diff
@@ -22,9 +22,17 @@ final public class Filter: NSObject {
...
         Use this to temporarily suspend examples that you know do not pass yet.
     */
+    public class var excluded: String {
+        return "excluded"
+    }
+
+    /**
+     Example and example groups with [Pending: true] are pending from test runs.
+     Use this to temporarily suspend examples that you know do not pass yet.
+     */
```
Comment (modocache) [inline Sources/Quick/Filter.swift]: nit-pick: the indentation for the documentation here doesn't match the indentation in the rest of this repository. It should be formatted like this:

```swift
/**
    Example and example groups with [Pending: true] are pending from test runs.
    Use this to temporarily suspend examples that you know do not pass yet.
*/
```

Comment (modocache) [inline Sources/Quick/Filter.swift]: nit-pick: I'd rephrase the documentation here:

```
Like "excluded", example and example groups with [Pending: true] are not included in test runs,
and so you can use this to temporarily suspend examples that you know do not pass yet.
"Pending" examples imply that the examples will soon pass, they just need a little more work.
Functionally, they are no different from "excluded" examples.
```

File: Sources/Quick/Filter.swift [hunk 5]
```diff
@@ -22,9 +22,17 @@ final public class Filter: NSObject {
     }
 
     /**
-        Example and example groups with [Pending: true] are excluded from test runs.
+        Example and example groups with [Excluded: true] are excluded from test runs.
         Use this to temporarily suspend examples that you know do not pass yet.
     */
```
Comment (modocache) [inline Sources/Quick/Filter.swift]: nit-pick: This documentation applies more to "pending" examples than "excluded" ones. I'd rephrase slightly:

```
Example and example groups with [Excluded: true] are not included in test runs.
Use this to disable examples that do not pass. Of course, you could also comment
the examples out altogether, but excluding them will ensure they continue to at
least compile, which allows them to be re-enabled more easily in the future.

In addition, future versions of Quick may be able to point out when a disabled test
passes, and thus should be re-enabled.
```

**Comment by modocache** [comment]
This looks great! I think the tests need some additional work, though. Thanks, @takecian!

**Comment by takecian** [comment]
Thank you for review! I'll fix it!

**Comment by takecian** [comment]
Hi @modocache , I've updated comment and added test for pending.
It would be helpful to review this.

Thanks!

File: Sources/QuickTests/FunctionalTests/PendingTests.swift [hunk 6]
```diff
@@ -5,24 +5,35 @@ import Nimble
...
 class FunctionalTests_PendingSpec: QuickSpec {
     override func spec() {
-        xit("an example that will not run") {
-            expect(true).to(beFalsy())
+        describe("a describe block containing only pending examples") {
+            beforeSuite { isBeforeSuiteCalled = true }
+            beforeEach { isBeforeEachCalled = true }
+            pending("an example that will not run") {}
+            afterEach { isAfterEachCalled = true }
+            afterSuite { isAfterSuiteCalled = true }
```
Comment (modocache) [inline Sources/QuickTests/FunctionalTests/PendingTests.swift]: So, this is *almost* what we want. However, the `beforeSuite` and `afterSuite` tests here will pass even without your changes. This is because this test suite contains non-pending examples. To really test these, you'd need an entirely new test target in the Quick workspace -- just like the QuickFunctionalTests target.

I think we should test this in a new QuickPendingTests target.

Comment (takecian) [inline Sources/QuickTests/FunctionalTests/PendingTests.swift]: Oh! I understand. I will create new test target and move this test into it.

**Comment by takecian** [comment]
I've created new test target for "pending" and merged master branch.
How do you feel?

**Comment by modocache** [comment]
Looking good! There are build failures on Linux, though. I think you'll need to use `#file` and `#line` instead of `__FILE__` and `__LINE__`.

File: Documentation/en-us/TestUsingTestDoubles.md [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+            
+            expect(mockProvier.fetchCalled).to(equal(true))
+        }
+    }
+}
+```
+
+Mock makes us easy to verify behavior which works with other objects.
+
+For more details about writing test, see https://realm.io/news/testing-in-swift/ .
```
Comment (modocache) [inline Documentation/en-us/TestUsingTestDoubles.md]: I am planning on reviewing this additional documentation in #487, but I don't think you meant to include it in this pull request. Could you take it out?

File: Documentation/ja/README.md [hunk 8]
```diff
@@ -30,6 +30,8 @@ Quickでテストを書くと、SwiftとObjective-Cで書かれたプログラ
   これでより簡潔にたくさんのテストが書けるようになります。
 - **[OS XとiOSアプリのテスト](TestingApps.md)**:
   AppKitとUIKitを使ったコードをどうやってテストできるか説明します。
+- **[Test doublesを使ったテスト](TestUsingTestDoubles.md)**:
+  Test doublesを使って対象のクラスのみをテストする方法を説明します。
```
Comment (modocache) [inline Documentation/ja/README.md]: Same as above, I think this doesn't belong in this pull request.

File: Documentation/ja/TestUsingTestDoubles.md [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+        }
+    }
+}
+```
+
+このようにオブジェクトのモックを作ることで動作をテストしやすくなります。
+
+テストの書き方について、更に詳細を知りたい方はこちらのビデオを参考にしてください。 https://realm.io/jp/news/testing-in-swift/ 。
+
+
```
Comment (modocache) [inline Documentation/ja/TestUsingTestDoubles.md]: Same as above. :)

File: Externals/Nimble [hunk 10]
```diff
@@ -1 +1 @@
-Subproject commit 0cf12916a276ba7bc7a66ded95bd8f7e77ca7f6b
+Subproject commit f24eac653b32a20b6c0f1d92c964ddd1adadfffc
```
Comment (modocache) [inline Externals/Nimble]: I think this might also be unnecessary. I don't think the pending changes would require an update to the version of Nimble being used...?

File: Sources/QuickTests/FunctionalTests/PendingTests.swift [hunk 11]
```diff
@@ -1,57 +0,0 @@
...
-        XCTAssertEqual(oneExampleBeforeEachExecutedCount, 1)
-    }
-
-    func testBeforeEachDoesNotRunForContextsWithOnlyPendingExamples() {
-        onlyPendingExamplesBeforeEachExecutedCount = 0
-
-        qck_runSpec(FunctionalTests_PendingSpec.self)
-        XCTAssertEqual(onlyPendingExamplesBeforeEachExecutedCount, 0)
-    }
-}
```
Comment (modocache) [inline Sources/QuickTests/FunctionalTests/PendingTests.swift]: Ah, sorry! I don't think these pending tests need to be *completely* removed from the main Quick test target. I just meant that, to test `beforeSuite` and `afterSuite` behavior for a test suite composed of **only pending tests**, we'd want a new test target. The `beforeEach` and `afterEach` tests should stay in this file.

Comment (takecian) [inline Sources/QuickTests/FunctionalTests/PendingTests.swift]: Oh, ok. I'll restore this source code!

**Comment by takecian** [comment]
[WIP] 
I did,
- Remove markdown file change of #487 
- Fix Nimble version
- Remove extra space
- Restore pending test suite for QuickFunctionalTest.

However build error on linux still occcurs. I will work for it.

**Comment by takecian** [comment]
@modocache , I've fixed errors on linux.
Could you review again?

File: Sources/QuickTests/FunctionalTests/PendingTests.swift (hunk 1 above)
Comment (takecian) [inline Sources/QuickTests/FunctionalTests/PendingTests.swift]: I updated test(add tests for `beforeEach`, `afterEach`, `beforeSuite`, and `afterSuite`) and create new target. 
See also, https://github.com/Quick/Quick/pull/491#discussion_r64206987 😄

**Comment by istx25** [comment]
@modocache: Do you have time to review this?

**Comment by istx25** [comment]
@modocache: Re-pinging, just in case the notification or email got lost. 👍

**Comment by modocache** [comment]
Didn't get lost, but it's a lot to review! 😅 

I'll need to find some time to read through it. Hopefully sometime this week.

**Comment by istx25** [comment]
Okay, no problem. Didn't know if you had seen it or not. 😸 I'll assign the PR to you?

**Comment by istx25** [comment]
Bringing attention back to this. Could we get the merge conflicts fixed @takecian and have it reviewed @modocache. Cheers! :)

**Comment by istx25** [comment]
@takecian: Please fix the merge conflicts so we can finish reviewing your changes.

**Comment by ikesyo** [comment]
Closing as this is stale/abandoned.
